sound/soc/sof/ipc.c

```diff
@@ -678,6 +688,138 @@ static void sof_set_get_mmio(struct snd_sof_dev *sdev,
 				       cdata->chanv, send_bytes);
 }
 
+static int sof_get_ctrl_size_params(struct snd_sof_control *scontrol,
+				    struct sof_ipc_ctrl_data *cdata,
+				    struct sof_ipc_ctrl_data_params *sparams)
+{
+	/*
+	 * Calculate the ctrl data offsets and sizes, because we have a union
+	 * as the last member and the scontrol->size doesn't reflect the actual
+	 * payload size, but rather the maximum.
+	 */
+
+	switch (cdata->type) {
+	case SOF_CTRL_TYPE_VALUE_CHAN_GET:
+	case SOF_CTRL_TYPE_VALUE_CHAN_SET:
+		sparams->msg_bytes = scontrol->num_channels *
+			sizeof(struct sof_ipc_ctrl_value_chan);
+		sparams->hdr_bytes = sizeof(struct sof_ipc_ctrl_data);
+		sparams->elems = scontrol->num_channels;
+		break;
+	case SOF_CTRL_TYPE_VALUE_COMP_GET:
+	case SOF_CTRL_TYPE_VALUE_COMP_SET:
+		sparams->msg_bytes = scontrol->num_channels *
+			sizeof(struct sof_ipc_ctrl_value_comp);
+		sparams->hdr_bytes = sizeof(struct sof_ipc_ctrl_data);
+		sparams->elems = scontrol->num_channels;
+		break;
+	case SOF_CTRL_TYPE_DATA_GET:
+	case SOF_CTRL_TYPE_DATA_SET:
+		sparams->msg_bytes = cdata->data->size;
+		sparams->hdr_bytes = sizeof(struct sof_ipc_ctrl_data) +
+			sizeof(struct sof_abi_hdr);
+		sparams->elems = cdata->data->size;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	cdata->rhdr.hdr.size = sparams->msg_bytes + sparams->hdr_bytes;
+	cdata->num_elems = sparams->elems;
+	cdata->total_elems = sparams->elems;
+
+	return 0;
+}
+
+static int sof_get_ctrl_copy_params(enum sof_ipc_ctrl_type ctrl_type,
+				    struct sof_ipc_ctrl_data *src,
+				    struct sof_ipc_ctrl_data *dst,
+				    struct sof_ipc_ctrl_data_params *sparams)
+{
+	switch (ctrl_type) {
+	case SOF_CTRL_TYPE_VALUE_CHAN_GET:
+	case SOF_CTRL_TYPE_VALUE_CHAN_SET:
+		sparams->src = (unsigned char *)src->chanv;
+		sparams->dst = (unsigned char *)dst->chanv;
```

Comment:
these casts and others below should be "(u8 *)"

Reply:
ok